UBL_DELTA post merge cleanup

oldmcg opened the pull request:
- fix fade_height, lost during some previous merge
- ~~fix float cx,cy which are not const~~
- move repeated z_cxcy calc line inside loop
- style fixes and comment fixes/alignment
- add UBL_DELTA_ON_CARTESIAN to allow finer grained UBL segmentation on cartesians

oldmcg committed:
UBL_DELTA post merge cleanup:
   fix fade_height, lost during some previous merge
   fix float cx,cy which are not const
   move repeated z_cxcy calc line inside loop
   style fixes and comment fixes/alignment

thinkyhead commented on Marlin/Conditionals_post.h:
```diff
@@ -731,7 +731,7 @@
    */
 
-  #define UBL_DELTA  (ENABLED(AUTO_BED_LEVELING_UBL) && ENABLED(DELTA))
+  #define UBL_DELTA  (ENABLED(AUTO_BED_LEVELING_UBL) && (ENABLED(DELTA) || ENABLED(UBL_DELTA_ON_CARTESIAN)))
```
If `UBL_DELTA` no longer refers to something delta-specific, then a new name is needed. Meanwhile, `UBL_DELTA_ON_CARTESIAN` is not a good name. Please change. Delta is delta and cartesian is cartesian. Don't "brand name" parts of UBL with a "delta" moniker, if they are not delta-specific.

If the aim is to have all UBL functions working on all systems, like `AUTO_BED_LEVELING_BILINEAR` now does, then we should finish the job and just get rid of all these conditionals.

oldmcg replied:
There are now two different UBL line routines that work on cartesian.  Roxy's original version, which segments on mesh cell boundaries.  And the new "delta" routine that segments on granular boundaries (0.1mm default on delta, 1.0mm default on cartesian).  For a cartesian machine, the "delta" UBL line routine will "hug" the mesh within the cell instead of jumping from one cell wall intercept height to another, but probably at the cost of lower maximum feedrate.  The "current" UBL cartesian routine, Roxy's routine, should be the default for cartesian.  But there needs to be some flag to switch to the "delta" routine or "granular" routine or "segmented" routine used by deltas so cartesian can TRY it.  The name I chose is not the permanent name, and not advertised in config.h.  I just need something to let people try it on cartesians.

oldmcg replied:
OK, I changed the name from UBL_DELTA_ON_CARTESIAN which is confusing to UBL_GRANULAR_SEGMENTATION_FOR_CARTESIAN to better reflect its meaning.  It is NOT the default for UBL on cartesian, just an option.

thinkyhead commented on Marlin/ubl_motion.cpp:
```diff
@@ -639,13 +643,14 @@
                     z_xmy1 = (z_x1y1 - z_x0y1) * (1.0 / (MESH_X_DIST));   // z slope per x along y1 (upper left to upper right)
 
-              float z_cxy0 = z_x0y0 + z_xmy0 * cx;          // z height along y0 at cx
+              float z_cxy0 = z_x0y0 + (z_xmy0 * cx);          // z height along y0 at cx
```
Ok, but not needed. Less pedantic code is preferred.
(i.e., If the compiler gets it, it's fine.)

oldmcg replied:
You had made other edits to my code that ADDED parentheses to expression like this, so I was trying to be consistent.  I preferred the original way.   Which way do you want it?

oldmcg replied:
Never mind -- my mistake, you had removed some extra parenthesis and I read the merge backward.

oldmcg replied:
Ok, the superfluous parentheses have again been expunged.

thinkyhead reviewed (changes requested):
- Please change `UBL_DELTA_ON_CARTESIAN`.
- Make UBL work everywhere and we can just call it "UBL" again.

thinkyhead commented on Marlin/ubl_motion.cpp:
```diff
@@ -589,6 +589,10 @@
 
+      #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
+        const float fade_scaling_factor = ubl.fade_scaling_factor_for_z(ltarget[Z_AXIS]);
+      #endif
```
I removed these lines because the line below (`z_cxcy *= fade_scaling_factor`) was not present, thus `fade_scaling_factor` was unused. Glad it is now restored!

oldmcg replied:
Yes, and that is how I noticed the other half of the fade semantics had disappeared during one of my many git merge/rebase sessions. So thanks for removing it!  That's how I noticed something was wrong.

oldmcg committed:
Update ubl_motion.cpp

remove unnecessary parentheses

oldmcg committed:
Update Conditionals_post.h

Change name of define to more accurate meaning:
UBL_GRANULAR_SEGMENTATION_FOR_CARTESIAN
which is not and should not be the default for cartesians with UBL.

oldmcg commented:
So where do we go from here?  Roxy separately fixed the incorrect const qualifiers that prevented it from building.  I would still like to fix fade height at a minimum.

thinkyhead commented:
Let me take a look at what we have here and see if it can be resolved easily. Typically, I would just start afresh at this point, or if the branch was believed to be in reasonable shape, take the time first to rebase it onto the most recent commits.

thinkyhead committed:
Merge branch 'bugfix-1.1.x' into ubl_delta

thinkyhead commented:
It was simple enough to resolve with the Github "resolve" button.

thinkyhead commented:
I'm going to give this "squash and merge" thing a try once this passes Travis CI….

thinkyhead commented:
So… While I think "Granular segmentation for cartesian" is a great idea, I think we should go full hog and implement it as `MOVE_SEGMENTATION` — period. We should enable it always for `DELTA` and `SCARA` (`IS_KINEMATIC`), and we should (eventually) enable it on cartesian when any mesh-based leveling system is active. (Maybe just as an option.) For Delta, it should have the segments-per-second option to decide how many segments to generate. For cartesian and SCARA we don't necessarily need to have the segments-per-second stuff.

What I'm describing is a generalization of the movement system, so we can include the move segmentation step based on just one conditional. Segmentation should also switch itself off on cartesian when leveling is disabled, using normal long movements to save on CPU.

That is essentially the level of simplification employed in Smoothieware, as I've lately discovered. For cartesian we would use a longer segment length than on a kinematic system. Though, if segments are small enough, then there will be no more need to do the split on grid boundaries step, saving on that bit of processing.

Segmentation for cartesian would also remove the need for the bilinear subdivision code, which calculates a finer (and SRAM costly) grid by a similar interpolation method to straight bilinear.

thinkyhead commented on Marlin/Conditionals_post.h:
```diff
@@ -731,7 +731,7 @@
    */
 
-  #define UBL_DELTA  (ENABLED(AUTO_BED_LEVELING_UBL) && ENABLED(DELTA))
+  #define UBL_DELTA  (ENABLED(AUTO_BED_LEVELING_UBL) && (ENABLED(DELTA) || ENABLED(UBL_GRANULAR_SEGMENTATION_FOR_CARTESIAN)))
```
So… by extending our repertoire of general flags with this notion…
```cpp
#define CARTESIAN_SEGMENTATION ENABLED(UBL_CARTESIAN_SEGMENTATION)
#define MOVE_SEGMENTATION (IS_KINEMATIC || CARTESIAN_SEGMENTATION)
```
…We will then have a simple conditional that we can use to decide whether to include and to call the `prepare_kinematic_move_to` (rename to `prepare_segmented_move_to`?) function which splits up moves for a kinematic system, or whether to call the normal movement stuff — which should still preserve the split-on-mesh-boundaries code.

And, maybe `CARTESIAN_SEGMENTATION` eventually becomes a general MBL/bilinear/UBL leveling option instead of a conditional. But for now… until all mesh leveling systems support it, this is good…

thinkyhead commented:
Ok… merging this now… but let's continue to look at making move-splitting a general option for all.

oldmcg commented:
Regarding the names and hierarchy of the #defines for now, I don't care as long as there is some way to turn on the ubl_prepare_linear_move_to routine as standard for UBL+DELTA and UBL+SCARA and as optional for UBL+CARTESIAN.

Up to the point of getting UBL for DELTA functionally into the codebase, I was trying to follow the existing model and avoid refactoring code for fear of introducing new bugs unrelated to specifically UBL_DELTA.  But there are definitely some opportunities for improvement and aggregation in how the code is layered for kinematic, segmentation, leveling, raw, logical, line, arc, etc, with the various requirements of each.

Roxy-3D commented:
> and as optional for UBL+CARTESIAN.

Yeah!   I spent a week re-writing the current ubl_motion.cpp file.  But I saw what you did, and I like it a lot!   For sure, we want to let Cartesian users turn on your code and use it!   And after we have some test time on it...  We can do a hard switch to it.
